lib/src/main/java/graphql/nadel/engine/transform/hydration/NadelHydrationTransform.kt

```diff
@@ -154,30 +166,104 @@ internal class NadelHydrationTransform(
             flatten = true,
         )
 
-        val jobs: List<Deferred<List<NadelResultInstruction>>> = coroutineScope {
-            parentNodes.map {
-                async {
-                    hydrate(
+        return if (isDeferred(executionContext, executionBlueprint, overallField)) {
+            deferHydration(parentNodes, state, executionBlueprint, overallField, executionContext)
+            return emptyList()
+        } else {
+            getResultInstructions(parentNodes, state, executionBlueprint, overallField, executionContext)
+        }
+    }
+
+    private suspend fun getResultInstructions(
+        parentNodes: List<JsonNode>,
+        state: State,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        overallField: ExecutableNormalizedField,
+        executionContext: NadelExecutionContext,
+    ): List<NadelResultInstruction> {
+        return coroutineScope {
+            parentNodes
+                .map {
+                    prepareHydration(
```

Comment:
So the way this works is we now split the hydration into two steps

1. Preparing the hydration
2. Executing the prepared hydration

The first step involves gathering things like the source inputs etc. This is done before the Nadel execution framework removes all the artificial fields etc.

Then for non-deferred hydration we execute step 2 in sync. Otherwise if it's deferred, we execute step 2 in a background coroutine and then build a delayed result and send it back to the user.